### ISSUE TITLE
ES6 destructuring with `this` variable

### ISSUE THREAD

**Issue by Guardiannw** [Microsoft/TypeScript#10076]
**TypeScript Version:**  1.8.10
**Code**

```ts
// A *self-contained* demonstration of the problem follows...
var array = [1, 2, 3];
[this.test, this.test2, this.test3] = array;
```

**Expected behavior:**
for it to compile as `_this.test = array[0], _this.test2 = array[1], _this.test3 = array[2];`
**Actual behavior:** 
it compiles as `_this.test = array[0], this.test2 = array[1], this.test3 = array[2];`

This is a problem because the `this` scope does not match correctly that of the lambda.

**Comment by RyanCavanaugh** [comment]
Can't repro on the Playground this with the code given. Can you post a complete sample (with the surrounding function) ?

**Comment by Guardiannw** [comment]
@RyanCavanaugh I could not reproduce this on the playground either.  But strangely i figured out a fix.

When I would write something to the effect of 
```
wrapper((array: [any]) => {
    [this.test, this.test1, this.test2] = array;
});
```
it would fail and would output as I had previously stated.
But when I changed the type, it worked correctly ->
```
wrapper((array: Array<any>) => {
    [this.test, this.test1, this.test2] = array;
});
```
I am not sure why this is.  I thought the type did not change transpilation, but it looks like it matters for that too.

**Comment by RyanCavanaugh** [comment]
Definitely a bug
```ts
declare function wrapper(x: any);
wrapper((array: [any]) => {
    [this.test, this.test1, this.test2] = array;
});
```
should output three `_this`s

**Comment by yuit** [comment]
@Guardiannw the reason why 
this example 
```typescript
wrapper((array: [any]) => {
    [this.test, this.test1, this.test2] = array;
});
```
and this one
```typescript
wrapper((array: Array<any>) => {
    [this.test, this.test1, this.test2] = array;
});
```
behaves differently is that the first example `array` has type tuple with one element (that is type `any`); when compiler type-check destructing assignment for the second element in tuple, it couldn't get type for that element so `this` expression is properly check (hence the flag to determine whether to capture lexically this doesn't get check as well). The second example is just an array so it doesn't affect by the similar issue

**Comment by Guardiannw** [comment]
@yuit That makes complete sense, but I thought that the types did not create any changes in the exported code, that they only made compile time type checks.  Is this correct or incorrect with the desired specs?

**Comment by yuit** [comment]
@Guardiannw I think this is incorrect and undesirable that we skip checking `this` expression and doesn't set the flag properly. Emitted code of both cases should be the same although one will give you a compiler error.

 I have merged the fix into `release-2.0` and `master` if you would like to give it a try!
